File: WordBreakRecursion.java
package DP5;

import java.util.HashSet;
import java.util.List;

/*
        -------------------------------------------------------------------------------------------------------
        Time complexity : o(n power n)
        space complexity: o(n power n)
        Did this code run successfully in leetcode : yes
        problems faces : no
 */
    
public class WordBreakRecursion {
    
public boolean wordBreak(String s, List<String> wordDict) {
    
    if(s == null || s.length() ==0 ) return false;
    
    HashSet<String> set = new HashSet<String>(wordDict);
    return  helper(s,set);
    
    }

private boolean helper(String s, HashSet<String> set) {
    
    //base
    if(s.length() == 0) return true;
    
    //logic
    for(int i=1;i<s.length();i++)
    {
        if(set.contains(s.substring(0, i)) && helper(s.substring(i),set) )
        {
            return true;
        }
    }
    
    return false;
}

}

File: word_break_recursion.py
# Time Complexity : O(2 ^ N) where N is length of the string
# Space Complexity : O(N) recursion stack depth in the worst case where N is length of the string
# Did this code successfully run on Leetcode : No (Time Limit Exceeded)
# Any problem you faced while coding this : No

class Solution:
    def __init__(self):
        self.wordsSet = set()
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        self.wordsSet = set(wordDict)
        return self.breaker(s)
    
    def breaker(self, s):
        #base
        if len(s) == 0: return True
        
        #logic
        for i in range(len(s)):
            if s[:i+1] in self.wordsSet and self.breaker(s[i+1:]):
                return True
            
        return False